File: packages/core/src/updates.ts
import { manifestContentHash, resolvedContentHash } from './hashing.js';
import type { GitPort } from './ports.js';
import type { FsPort } from './ports.js';
import type { InstallManifest, Repository, ResolvedSkill } from './model.js';

/** Revision used to resolve the tracked upstream ref. */
const UPSTREAM = '@{upstream}';

/**
 * Repository-level update detection. Fetches, then compares the local `HEAD`
 * to the tracked upstream ref. The repository "can be updated" when they differ.
 * Read-only: a fetch does not modify the working tree or any install.
 *
 * @param git Git port for the local, instant comparisons (rev-parse).
 * @param repo Repository whose `localPath` is checked.
 * @param fetchGit Git port used for the network fetch; defaults to `git`. Pass a
 *   terminal-backed port so the fetch runs in the interactive shell (visible,
 *   ssh-capable) like a pull, while the rev-parse comparisons stay silent.
 */
export async function repoHasUpdate(
  git: GitPort,
  repo: Repository,
  fetchGit: GitPort = git,
): Promise<boolean> {
  await fetchGit.fetch(repo.localPath);
  const local = await git.revParse(repo.localPath, 'HEAD');
  const upstream = await git.revParse(repo.localPath, UPSTREAM);
  return local.oid !== upstream.oid;
}

/**
 * Skill-level update detection. Compares the content hashes of the resolved
 * skill's body files in the working tree against the hashes recorded in the
 * install manifest. The skill "can be updated" when the source content differs
 * from what is installed (a changed, added, or removed file).
 *
 * @param fs Filesystem port.
 * @param sourceRoot Working-tree root the resolved paths are relative to.
 * @param resolved The freshly resolved skill from the working tree.
 * @param manifest The recorded install manifest.
 */
export async function skillHasUpdate(
  fs: FsPort,
  sourceRoot: string,
  resolved: ResolvedSkill,
  manifest: InstallManifest,
): Promise<boolean> {
  const source = await resolvedContentHash(fs, sourceRoot, resolved);
  const installed = manifest.contentHash ?? manifestContentHash(manifest);
  return source !== installed;
}

File: crates/skillkeeper-cli/src/updates.rs
//! Read-only update detection (port of `packages/core/src/state/updates.ts`).
//!
//! Divergence: the Rust domain crate `skillkeeper-core` has not (yet) ported
//! `state/updates.ts`, so the two helpers the `check` command needs live here in
//! the CLI. They are thin compositions over the already-ported git port and the
//! content-hashing helpers, so this is a location divergence, not a logic one.

use skillkeeper_core::hashing::{manifest_content_hash, resolved_content_hash};
use skillkeeper_core::models::{InstallManifest, Repository, ResolvedSkill};
use skillkeeper_core::ports::{FsPort, GitPort, PortResult};

/// Revision used to resolve the tracked upstream ref.
const UPSTREAM: &str = "@{upstream}";

/// Repository-level update detection. Fetches, then compares the local `HEAD`
/// to the tracked upstream ref. The repository can be updated when they differ.
/// Read-only: a fetch does not modify the working tree or any install.
pub fn repo_has_update(git: &dyn GitPort, repo: &Repository) -> PortResult<bool> {
    git.fetch(&repo.local_path)?;
    let local = git.rev_parse(&repo.local_path, "HEAD")?;
    let upstream = git.rev_parse(&repo.local_path, UPSTREAM)?;
    Ok(local.oid != upstream.oid)
}

/// Skill-level update detection. Compares the content hash of the resolved
/// skill's body files in the working tree against the hash recorded in the
/// install manifest. The skill can be updated when the source content differs
/// from what is installed.
pub fn skill_has_update(
    fs: &dyn FsPort,
    source_root: &str,
    resolved: &ResolvedSkill,
    manifest: &InstallManifest,
) -> PortResult<bool> {
    let source = resolved_content_hash(fs, source_root, resolved)?;
    let installed = manifest
        .content_hash
        .clone()
        .unwrap_or_else(|| manifest_content_hash(manifest));
    Ok(source != installed)
}
